doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+one we have in the `UPSERT` state.
+
+However, this means if we receive multiple requests with the same idempotency
+token, we will retract the initial state and leave the last-seen value in the
+persist shard. Because we plan have add the time the request was received to the
+row of data we append, this means we will continually update the time the
+request was written and extend the lifetime of the idempotency key from the last
+time the write occurred.
+
+I think this is reasonable, but wanted to make sure this behavior's is
+explicitly understood as a tradeoff of storing the time.
+
+**Alternative**
+
+It's possible to achieve a similar outcome by increasing the compaction period
+for the remap shard such that we can get the time at which each event occurred
+in Materialize time--this feedback loop seems challenging to instrument,
+however.
+
+### Architectural overview
+
+#### Dataflow
+
+##### Source rendering
+
+The rendered dataflow will simply read updates out of a channel and give them to
+a data capability at some cadence, downgrading on either some interval's tick or
+when the amount of buffered data reaches some threshold.
+
+Providing users a threshold of requests to buffer offers some knob to control
+throughput vs. latency for the source.
+
+**Timestamping**
+
+The native timestamp of the source could just be a `u64`, which we increment for
+each batch we produce.
+
+The annoying corner of this, though, is that the latency of any request is
+limited by the frequency with which we reclock batches. The most feasible option
+I see here is providing users the option to reclock on a shorter period. The
+current period is 1s, but it could be set to some configurable duration. I'm not
+sure how low this could be set to produce a meaningful effect for the user.
+
+**Request responses**
+
+To provide durability guarantees for webhook sources, we cannot respond to the
+client with a `200` message until we are sure that the request has been
+committed to the shard.
+
+We can retain an in-memory map of `BTreeMap<u64, BTreeSet<IdempotencyKey>>`, and
+then as we get feedback that a batch has been committed can send back all
+idempotency keys committed along response edge.
+
+This has the side-effect of forcing us to send `200` responses on writes whose
+idempotency keys we've already seen because the entire batch commits, i.e. we do
+not have the ability in the rendering to understand if a write committed _now_
+or was already present. I don't believe this to be problematic--a `500` response
+followed by a `200` response, irrespective of the "underlying" result of the
+`500` response still unambiguously states that the write committed.
+
+##### Upsert state reduction
+
+Currently, the `UPSERT` envelope purposefully retains all state for all time.
+However, that is _not_ desirable for a webhook source, whose domain of keys is
+unreasonably large.
+
+_I have not meaningfully dug into the feasibility of the following approach, so
```

Comment:
I would strongly recommend looking into RocksDB compaction filters for the RocksDB based upsert operator

Reply:
TIL––I hadn't realized RocksDB supported "expiring" its own keys. Very cool––ty for the pointer, Seth.